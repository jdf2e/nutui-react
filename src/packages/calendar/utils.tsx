import { Utils } from '@/utils/date'
import { CalendarDay, CalendarMonthInfo } from './types'

export const splitDate = (date: string) => {
  const split = date.indexOf('-') !== -1 ? '-' : '/'
  return date.split(split)
}

export const isMultiple = (day: string, days: string[]) => {
  if (days.length > 0) {
    return days.some((item: string) => {
      return Utils.isEqual(item, day)
    })
  }
  return false
}

export const isCurrDay = (month: CalendarMonthInfo, day: string | number) => {
  const date = `${month.curData[0]}/${month.curData[1]}/${day}`
  return Utils.isEqual(date, Utils.date2Str(new Date(), '/'))
}

export const getCurrDate = (day: CalendarDay, month: CalendarMonthInfo) => {
  return `${month.curData[0]}/${month.curData[1]}/${Utils.getNumTwoBit(
    +day.day
  )}`
}

export const isStart = (day: string, days: string[]) => {
  return Utils.isEqual(days[0], day)
}

export const isEnd = (day: string, days: string[]) => {
  return Utils.isEqual(days[1], day)
}

// 开始结束时间是否相等
export const isStartAndEnd = (days: string[]) => {
  return days.length >= 2 && Utils.isEqual(days[0], days[1])
}
