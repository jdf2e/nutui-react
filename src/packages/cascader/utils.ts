import {
  CascaderOption,
  CascaderConfig,
  CascaderFormat,
  CascaderValue,
} from './types'

export const formatTree = (
  tree: CascaderOption[],
  parent: CascaderOption | null,
  config: CascaderConfig
): CascaderOption[] =>
  tree.map((node: any) => {
    const {
      value: valueKey = 'value',
      text: textKey = 'text',
      children: childrenKey = 'children',
    } = config
    const {
      [valueKey]: value,
      [textKey]: text,
      [childrenKey]: children,
      ...others
    } = node
    const newNode: CascaderOption = {
      loading: false,
      ...others,
      level: parent ? ((parent && parent.level) || 0) + 1 : 0,
      value,
      text,
      children,
      _parent: parent,
    }
    if (newNode.children && newNode.children.length) {
      newNode.children = formatTree(newNode.children, newNode, config)
    }
    return newNode
  })

export const eachTree = (
  tree: CascaderOption[],
  cb: (node: CascaderOption) => unknown
): void => {
  let i = 0
  let node: CascaderOption
  while ((node = tree[i++])) {
    if (cb(node) === true) {
      break
    }
    if (node.children && node.children.length) {
      eachTree(node.children, cb)
    }
  }
}

const defaultConvertConfig = {
  topId: null,
  idKey: 'id',
  pidKey: 'pid',
  sortKey: '',
}
export const convertListToOptions = (
  list: CascaderOption[],
  options: CascaderFormat
): CascaderOption[] => {
  const mergedOptions = {
    ...defaultConvertConfig,
    ...(options || {}),
  }
  const { topId, idKey, pidKey, sortKey } = mergedOptions
  let result: CascaderOption[] = []
  let map: any = {}
  list.forEach((node: any) => {
    node = { ...node }
    const { [idKey]: id, [pidKey]: pid } = node
    const children = (map[pid] = map[pid] || [])
    if (!result.length && pid === topId) {
      result = children
    }
    children.push(node)
    node.children = map[id] || (map[id] = [])
  })

  if (sortKey) {
    Object.keys(map).forEach((i) => {
      if (map[i].length > 1) {
        map[i].sort((a: any, b: any) => a[sortKey] - b[sortKey])
      }
    })
  }
  map = null
  return result
}

class Tree {
  nodes: CascaderOption[]

  readonly config: CascaderConfig

  constructor(nodes: CascaderOption[], config?: CascaderConfig) {
    this.config = {
      value: 'value',
      text: 'text',
      children: 'children',
      ...(config || {}),
    }
    this.nodes = formatTree(nodes, null, this.config)
  }

  updateChildren(nodes: CascaderOption[], parent: CascaderOption | null): void {
    if (!parent) {
      this.nodes = formatTree(nodes, null, this.config)
    } else {
      parent.children = formatTree(nodes, parent, this.config)
    }
  }

  // for test
  getNodeByValue(value: CascaderOption['value']): CascaderOption | void {
    let foundNode
    eachTree(this.nodes, (node: CascaderOption) => {
      if (node.value === value) {
        foundNode = node
        return true
      }
      return null
    })
    return foundNode
  }

  getPathNodesByValue(value: CascaderValue): CascaderOption[] {
    if (!value.length) {
      return []
    }

    const pathNodes = []
    let currentNodes: CascaderOption[] | void = this.nodes

    while (currentNodes && currentNodes.length) {
      const foundNode: CascaderOption | void = currentNodes.find(
        (node) => node.value === value[node.level as number]
      )

      if (!foundNode) {
        break
      }

      pathNodes.push(foundNode)
      currentNodes = foundNode.children
    }

    return pathNodes
  }

  // eslint-disable-next-line class-methods-use-this
  isLeaf = (node: CascaderOption, lazy: boolean): boolean => {
    const { leaf, children } = node
    const hasChildren = Array.isArray(children) && Boolean(children.length)
    return leaf || (!hasChildren && !lazy)
  }

  hasChildren = (node: CascaderOption, lazy: boolean): boolean => {
    const isLeaf = this.isLeaf(node, lazy)

    if (isLeaf) {
      return false
    }

    const { children } = node
    return Array.isArray(children) && Boolean(children.length)
  }
}

export default Tree
