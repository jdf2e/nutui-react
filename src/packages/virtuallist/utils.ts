import { PositionType, Data } from './type'

// 缓存列表初始化信息
const initPositinoCache = (reaItemSize: number, length = 0): PositionType[] => {
  let index = 0
  const positions: PositionType[] = Array(length)
  while (index < length) {
    positions[index] = {
      index,
      height: reaItemSize,
      width: reaItemSize,
      top: index * reaItemSize,
      bottom: (index + 1) * reaItemSize,
      left: index * reaItemSize,
      right: (index + 1) * reaItemSize,
    }
    index++
  }
  return positions
}
// 获取列表总高度
const getListTotalSize = (
  positions: Array<PositionType>,
  horizontal: true | false
): number => {
  const index = positions.length - 1
  let size = 0
  if (index < 0) {
    size = 0
  } else {
    size = horizontal ? positions[index].right : positions[index].bottom
  }
  return size
}
// 通过二分法找到 scrollOffset 对应的值
const binarySearch = (
  positionsList: Array<PositionType>,
  value = 0,
  horizontal: true | false
): number => {
  let start = 0
  let end: number = positionsList.length - 1
  let tempIndex = null
  const key = horizontal ? 'right' : 'bottom'
  while (start <= end) {
    const midIndex = Math.floor((start + end) / 2)
    const midValue = positionsList[midIndex][key]

    // 相等则直接返回（因为是bottom, 因此startIndex应该是下一个节点）
    if (midValue === value) {
      return midIndex + 1
    }
    // 中间值 < 传入值，则说明 value对应的节点 大于 start, start往后移动一位
    if (midValue < value) {
      start = midIndex + 1
    }
    // 中间值 > 传入值，则说明 value 在 中间值之前，end 节点移动到 mid - 1
    else if (midValue > value) {
      // tempIndex存放最靠近值为value的所有
      if (tempIndex === null || tempIndex > midIndex) {
        tempIndex = midIndex
      }
      end = midIndex - 1
    }
  }
  tempIndex = tempIndex || 0
  return tempIndex
}
const getEndIndex = ({
  sourceData,
  startIndex,
  visibleCount,
  itemEqualSize = true,
  positions,
  offSetSize,
  overscan,
  sizeKey = 'width',
}: {
  sourceData: Array<Data>
  startIndex: number
  visibleCount: number
  itemEqualSize?: boolean
  positions: PositionType[]
  offSetSize: number
  overscan: number
  sizeKey?: 'width' | 'height'
}): number => {
  const dataLength = sourceData.length
  let tempIndex = null
  if (itemEqualSize) {
    const endIndex = startIndex + visibleCount
    tempIndex = dataLength > 0 ? Math.min(dataLength, endIndex) : endIndex
  } else {
    let sizeNum = 0
    for (let i = startIndex; i < dataLength; i++) {
      sizeNum += positions[i][sizeKey] || 0
      if (sizeNum > offSetSize) {
        const endIndex = i + overscan
        tempIndex = dataLength > 0 ? Math.min(dataLength, endIndex) : endIndex
        break
      }
    }
    if (sizeNum < offSetSize) {
      tempIndex = dataLength
    }
  }
  tempIndex = tempIndex || 0
  return tempIndex
}

// 更新Item大小
const updateItemSize = (
  positions: PositionType[],
  items: HTMLCollection,
  sizeKey: 'width' | 'height'
): void => {
  const newPos = positions.concat()
  Array.from(items).forEach((item) => {
    const index = Number(item.getAttribute('data-index'))
    const styleVal = item.getAttribute('style')
    if (styleVal && styleVal.includes('none')) return
    const nowSize = item.getBoundingClientRect()[sizeKey]

    const oldSize = positions[index][sizeKey] as number
    // 存在差值, 更新该节点以后所有的节点
    const dValue = oldSize - nowSize
    if (dValue) {
      if (sizeKey === 'width') {
        newPos[index].right -= dValue
        newPos[index][sizeKey] = nowSize
        for (let k = index + 1; k < positions.length; k++) {
          newPos[k].left = positions[k - 1].right
          newPos[k].right -= dValue
        }
      } else if (sizeKey === 'height') {
        newPos[index].bottom -= dValue
        newPos[index][sizeKey] = nowSize
        for (let k = index + 1; k < positions.length; k++) {
          newPos[k].top = positions[k - 1].bottom
          newPos[k].bottom -= dValue
        }
      }
    }
  })
}

export {
  initPositinoCache,
  getListTotalSize,
  binarySearch,
  getEndIndex,
  updateItemSize,
}
