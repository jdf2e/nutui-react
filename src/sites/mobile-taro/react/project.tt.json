{
  "miniprogramRoot": "./",
  "projectname": "first",
  "appid": "testAppId",
  "setting": {
    "es6": false,
    "minified": false
  }
}
