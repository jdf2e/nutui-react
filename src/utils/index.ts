export const isObject = (val: unknown): val is Record<any, any> =>
  val !== null && typeof val === 'object'
