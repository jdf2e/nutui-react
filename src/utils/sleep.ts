export const sleep = (time: number): Promise<unknown> =>
  new Promise((resolve) => setTimeout(resolve, time))
