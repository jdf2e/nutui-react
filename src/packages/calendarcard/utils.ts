import { Utils } from '@/utils/date'
import { CalendarDay } from './type'

export const convertDateToDay = (date: Date) => {
  return date
    ? {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
        date: date.getDate(),
      }
    : null
}

export const convertDayToDate = (day: CalendarDay) => {
  return day ? new Date(day.year, day.month - 1, day.date) : null
}

/**
 * 获取当月面板中前一个月的日期数据
 */
export const getPrevMonthDays = (
  year: number,
  month: number,
  firstDayOfWeek: number
) => {
  let prevMonth = month - 1
  let prevYear = year
  if (prevMonth <= 0) {
    prevMonth = 12
    prevYear -= 1
  }
  let days = Utils.getMonthPreDay(year, month)
  days -= firstDayOfWeek
  if (days >= 7) {
    days -= 7
  }

  const preDates = Utils.getMonthDays(`${prevYear}`, `${prevMonth}`)
  const months = Array.from(Array(preDates), (_, k) => {
    return {
      type: 'prev',
      year: prevYear,
      month: prevMonth,
      date: k + 1,
    } as CalendarDay
  })
  return months.slice(preDates - days)
}

/**
 * 获取当前月的日期数据
 */
export const getCurrentMonthDays = (year: number, month: number) => {
  const days = Utils.getMonthDays(`${year}`, `${month}`)
  return Array.from(Array(days), (_, k) => {
    return {
      type: 'current',
      year,
      month,
      date: k + 1,
    } as CalendarDay
  })
}

/**
 * 根据日期获取当前周的起始日期
 */
export const getCurrentWeekDays = (
  day: CalendarDay,
  firstDayOfWeek: number
): CalendarDay[] => {
  const current = new Date(day.year, day.month - 1, day.date)
  const count = (current.getDay() + 7 - firstDayOfWeek) % 7
  return [
    convertDateToDay(
      new Date(current.getTime() - 24 * 60 * 60 * 1000 * count)
    ) as CalendarDay,
    convertDateToDay(
      new Date(current.getTime() + 24 * 60 * 60 * 1000 * (6 - count))
    ) as CalendarDay,
  ]
}
