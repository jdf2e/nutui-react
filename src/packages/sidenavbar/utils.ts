import { MouseEventHandler } from 'react'

const handleClick: MouseEventHandler<HTMLDivElement> = (e) => {
  e.stopPropagation()
  const isIcon = (e.target as HTMLDivElement).className.includes('arrow-icon')
  const isTitle =
    (e.target as HTMLDivElement).className.includes('__title') || isIcon
  const currentClass = e.currentTarget.className
  const isShow = currentClass.includes('nutShow')
  const arrowIcon = e.currentTarget.querySelector('.arrow-icon') as Element
  const iconClass = arrowIcon.className

  if (isTitle) {
    e.currentTarget.className = isShow
      ? currentClass.replace('nutShow', 'nutHide')
      : currentClass.replace('nutHide', 'nutShow')
    arrowIcon.className = isShow
      ? iconClass.replace('arrow-down', 'arrow-up')
      : iconClass.replace('arrow-up', 'arrow-down')
  }
}
export { handleClick }
