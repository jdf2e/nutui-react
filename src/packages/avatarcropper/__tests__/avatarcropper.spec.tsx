import React from 'react'
import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import AvatarCropper, { AvatarCropperRef } from '../index'
import { simulateTouchMove, simulateTouchZoom } from '@/utils/test/event'
import { sleep } from '@/utils/sleep'
import Button from '@/packages/button'

const mockFile = new File([new ArrayBuffer(10000)], 'test.jpg', {
  type: 'image/jpg',
})

test('layout default slot', () => {
  const { container } = render(
    <AvatarCropper>
      <img
        src="https://img12.360buyimg.com/imagetools/jfs/t1/196430/38/8105/14329/60c806a4Ed506298a/e6de9fb7b8490f38.png"
        alt="avatar"
      />
    </AvatarCropper>
  )

  expect(screen.getByAltText('avatar')).toBeInTheDocument()
  expect(container.querySelector('.nut-avatar-cropper')).toMatchSnapshot()
})

test('should render base cutAvatar and type', async () => {
  const { container } = render(<AvatarCropper />)

  expect(container.querySelector('.nut-avatar-cropper')).toBeInTheDocument()

  const input = container.querySelector('.nut-avatar-cropper__input')
  expect(input?.getAttribute('type')).toBe('file')
})

test('layout toolbar slot', () => {
  const handleCancel = jest.fn()
  const handleConfirm = jest.fn()
  const avatarCropperRef = React.createRef<AvatarCropperRef>()
  const Toolbar = () => {
    return (
      <div className="toolbar">
        <Button
          type="primary"
          onClick={(e: any) => avatarCropperRef.current?.cancel()}
        >
          取消
        </Button>
        <Button
          type="primary"
          onClick={(e: any) => avatarCropperRef.current?.reset()}
        >
          重置
        </Button>
        <Button
          type="primary"
          onClick={(e: any) => avatarCropperRef.current?.rotate()}
        >
          旋转
        </Button>
        <Button
          type="primary"
          onClick={(e: any) => avatarCropperRef.current?.confirm()}
        >
          确认
        </Button>
      </div>
    )
  }
  const { container } = render(
    <AvatarCropper
      ref={avatarCropperRef}
      toolbar={Toolbar()}
      onCancel={handleCancel}
      onConfirm={handleConfirm}
    >
      <img
        src="https://img12.360buyimg.com/imagetools/jfs/t1/196430/38/8105/14329/60c806a4Ed506298a/e6de9fb7b8490f38.png"
        alt="avatar"
      />
    </AvatarCropper>
  )
  expect(container.querySelector('.toolbar')).toBeInTheDocument()
  // 验证toolbar的点击事件
  const toolbarItems = container.querySelectorAll('.toolbar button')
  expect(toolbarItems.length).toBe(4)
  const cancel = toolbarItems[0]
  fireEvent.click(cancel)
  expect(handleCancel).toBeCalled()
  const reset = toolbarItems[1]
  fireEvent.click(reset)
  const rotate = toolbarItems[2]
  fireEvent.click(rotate)
  fireEvent.click(rotate)
  fireEvent.click(rotate)
  fireEvent.click(rotate)
  const confirm = toolbarItems[3]
  fireEvent.click(confirm)
})

test('AvatarCropper: Select the image to open the crop window', async () => {
  const handleCancel = jest.fn()
  const handleConfirm = jest.fn()
  const { container } = render(
    <AvatarCropper
      maxZoom={2}
      space={20}
      toolbarPosition="top"
      editText="修改"
      cancelText="取消"
      confirmText="裁剪"
      onCancel={handleCancel}
      onConfirm={handleConfirm}
    />
  )

  const input = container.querySelector('.nut-avatar-cropper__input') as
    | Element
    | Node
    | Document
    | Window
  expect(input).toBeInTheDocument()

  const smallFile = new File([new ArrayBuffer(100)], 'small.jpg')

  Object.defineProperty(input, 'files', {
    get: jest.fn().mockReturnValue([mockFile, smallFile]),
  })

  expect(container.querySelector('.nut-cropper-popup')).toHaveStyle({
    display: 'none',
  })
  fireEvent.change(input)
  await sleep(10)
  expect(container.querySelector('.nut-cropper-popup')).toHaveStyle({
    display: 'block',
  })
  expect(
    container.querySelector('.nut-cropper-popup__canvas')
  ).toBeInTheDocument()

  const track = container.querySelector('.nut-cropper-popup__highlight')

  await simulateTouchZoom(track as HTMLElement, 0, 400, 1000)

  const toolbarItems = container.querySelectorAll(
    '.nut-cropper-popup__toolbar-item'
  )
  expect(toolbarItems.length).toBe(4)

  const cancel = toolbarItems[0]
  fireEvent.click(cancel)
  expect(container.querySelector('.nut-cropper-popup')).toHaveStyle({
    display: 'none',
  })
  expect(handleCancel).toBeCalled()

  await sleep(10)

  const reset = toolbarItems[1]
  fireEvent.click(reset)

  const rotate = toolbarItems[2]
  fireEvent.click(rotate)
  // 模拟触摸移动
  await simulateTouchMove(track as HTMLElement, 10, 10, 1000, 800, 1000)

  fireEvent.click(rotate)
  fireEvent.click(rotate)
  fireEvent.click(rotate)

  const confirm = toolbarItems[3]
  fireEvent.click(confirm)

  expect(container.querySelector('.nut-cropper-popup')).toHaveStyle({
    display: 'none',
  })
})
