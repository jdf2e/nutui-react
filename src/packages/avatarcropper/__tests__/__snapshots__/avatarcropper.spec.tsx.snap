// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout default slot 1`] = `
<div
  class="nut-avatar-cropper"
>
  <img
    alt="avatar"
    src="https://img12.360buyimg.com/imagetools/jfs/t1/196430/38/8105/14329/60c806a4Ed506298a/e6de9fb7b8490f38.png"
  />
  <input
    accept="image/*"
    aria-label="选择图片"
    class="nut-avatar-cropper-input"
    type="file"
  />
  <div
    class="nut-avatar-cropper-edit-text"
  >
    编辑
  </div>
</div>
`;
